lib/db/backend/badger_backend.go

```diff
@@ -0,0 +1,377 @@
+	return badgerPrefixIterator(l.txn, prefix), nil
+}
+
+func (l badgerSnapshot) NewRangeIterator(first, last []byte) (Iterator, error) {
+	return badgerRangeIterator(l.txn, first, last), nil
+}
+
+func (l badgerSnapshot) Release() {
+	defer l.rel.Release()
+	l.txn.Discard()
+}
+
+// badgerTransaction implements backend.WriteTransaction using a batch (not
+// an actual badger transaction)
+type badgerTransaction struct {
+	badgerSnapshot
+	txn  *badger.Txn
+	bdb  *badger.DB
+	rel  *releaser
+	size int
+}
+
+func (t *badgerTransaction) Delete(key []byte) error {
+	t.size += len(key)
+	kc := make([]byte, len(key))
+	copy(kc, key)
+	return t.transactionRetried(func(txn *badger.Txn) error {
+		return txn.Delete(kc)
+	})
+}
+
+func (t *badgerTransaction) Put(key, val []byte) error {
+	t.size += len(key) + len(val)
+	kc := make([]byte, len(key))
+	copy(kc, key)
+	vc := make([]byte, len(val))
+	copy(vc, val)
+	return t.transactionRetried(func(txn *badger.Txn) error {
+		return txn.Set(kc, vc)
+	})
+}
+
+// transactionRetried performs the given operation in the current
+// transaction, with commit and retry if Badger says the transaction has
+// grown too large.
+func (t *badgerTransaction) transactionRetried(fn func(*badger.Txn) error) error {
+	if err := fn(t.txn); err == badger.ErrTxnTooBig {
+		if err := t.txn.Commit(); err != nil {
+			return wrapBadgerErr(err)
+		}
+		t.size = 0
+		t.txn = t.bdb.NewTransaction(true)
+		return wrapBadgerErr(fn(t.txn))
+	} else if err != nil {
+		return wrapBadgerErr(err)
+	}
+	return nil
+}
+
+func (t *badgerTransaction) Commit() error {
+	defer t.rel.Release()
+	defer t.badgerSnapshot.Release()
+	return wrapBadgerErr(t.txn.Commit())
+}
+
+func (t *badgerTransaction) Checkpoint() error {
+	if t.size < checkpointFlushMinSize {
+		return nil
+	}
+	err := t.txn.Commit()
+	if err == nil {
+		t.size = 0
+		t.txn = t.bdb.NewTransaction(true)
+	}
+	return wrapBadgerErr(err)
+}
+
+func (t *badgerTransaction) Release() {
+	defer t.rel.Release()
+	defer t.badgerSnapshot.Release()
+	t.txn.Discard()
+}
+
+type badgerIterator struct {
+	it        *badger.Iterator
+	prefix    []byte
+	first     []byte
+	last      []byte
+	releaseFn func()
+	didSeek   bool
+}
+
+func (i *badgerIterator) Next() bool {
+	for {
+		if !i.didSeek {
+			i.it.Seek(i.prefix)
+			i.didSeek = true
+		} else {
+			i.it.Next()
+		}
+
+		if !i.it.ValidForPrefix(i.prefix) {
+			// Done
+			return false
+		}
+		if i.first == nil && i.last == nil {
+			// No range checks required
+			return true
+		}
+
+		key := i.it.Item().Key()
+		if bytes.Compare(key, i.first) < 0 {
+			// Key is before range first
+			continue
+		}
+		if bytes.Compare(key, i.last) > 0 {
+			// Key is after range last
+			return false
+		}
+		return true
+	}
+}
+
+func (i *badgerIterator) Key() []byte {
+	return i.it.Item().Key()
+}
+
+func (i *badgerIterator) Value() []byte {
+	val, err := i.it.Item().ValueCopy(nil)
+	if err != nil {
```

Comment:
Change the db api to reduce garbage generatwd while we're at it?

Reply:
Maybe when/if we deprecate leveldb? It seems premature and excessive to me right now